src/lib.rs

```diff
@@ -33,51 +36,50 @@ pub fn gini(data: &[u64]) -> f32 {
 /// Returns a float where 1.0 is a perfectly accurate dataset
 /// ```
 /// use parsnip::categorical_accuracy;
-/// let pred = vec![0, 0, 0 , 1, 2];
-/// let actual = vec![1, 1, 1, 1, 2];
+/// let pred : Vec<u16> = vec![0, 0, 0 , 1, 2];
+/// let actual : Vec<u16> = vec![1, 1, 1, 1, 2];
 /// assert_eq!(categorical_accuracy(&pred, &actual), 0.4);
 /// ```
-pub fn categorical_accuracy(pred: &[u64], actual: &[u64]) -> f32 {
+pub fn categorical_accuracy<T: Unsigned + ToPrimitive>(pred: &[T], actual: &[T]) -> f32 {
     assert_eq!(pred.len(), actual.len());
-    let bools =  pred.iter().zip(actual).map(|(x,y)| x == y);
-    let truthy : Vec<bool> =  bools.filter(|b| *b).collect();
-    truthy.len() as f32 / pred.len() as f32
+    let truthy =  pred.iter().zip(actual).filter(|(x,y)| x == y).count();
+    return truthy as f32 / pred.len() as f32;
 }
 
-fn class_precision(pred: &[u64], actual: &[u64], class: u64) -> f32 {
+fn class_precision<T: Unsigned + ToPrimitive>(pred: &[T], actual: &[T], class: T) -> f32 {
     assert_eq!(pred.len(), actual.len());
-    let true_positives_map = pred.iter().zip(actual).map(|(p, a)| p == a && *p == class);
-    let true_positives = true_positives_map.filter(|b| *b).count() as f32;
-    let all_positives = pred.iter().map(|p| *p == class).filter(|b| *b).count() as f32;
+    //First, get the map of all true positives
+    let true_positives = pred.iter().zip(actual).filter(|(p, a)| p == a && **p == class).count() as f32;
+    let all_positives = pred.iter().filter(|p| **p == class).count() as f32;
     if all_positives == 0.0 {
         0.0
     } else {
         true_positives / all_positives
     }
 }
 
-fn weighted_precision(pred: &[u64], actual: &[u64]) -> f32 {
+fn weighted_precision<T: Unsigned + ToPrimitive + Ord + Clone>(pred: &[T], actual: &[T]) -> f32 {
     assert_eq!(pred.len(), actual.len());
-    let mut classes : Vec<u64> = pred.into_iter().map(|x| *x).collect();
+    let mut classes : Vec<&T> = pred.into_iter().collect();
     let mut class_weights = HashMap::new();
     classes.sort();
     classes.dedup();
     for value in classes.clone() {
-        class_weights.insert(value, actual.iter().filter(|a| **a == value).count() as f32 / actual.len() as f32);
+        class_weights.insert(value.to_u128().unwrap(), actual.iter().filter(|a| *a == value).count() as f32 / actual.len() as f32);
     }
-    classes.iter().map(|c| class_precision(pred, actual, *c) * class_weights.get(c).unwrap()).sum()
+    return classes.iter().map(|c| class_precision(pred, actual, (**c).clone()) * class_weights.get(&c.to_u128().unwrap()).unwrap()).sum();
```

Comment:
Since `classes` will be dropped anyway at the end of the function, we can just take ownership of its elements using `classes.into_iter()` instead. More changes related to borrowing and dereferencing may take place, but will generally make the code simpler.

Moreover, using the `Index` operator (`a[b]`) does the same thing as `a.get(b).unwrap()`:

```
class_weights[&c.to_usize().unwrap()]
```